### PR TITLE
OCDEV e2e tests on travis:

### PR DESCRIPTION
**Pull request by ashetty1**
1. Pushing a separate PR with glide changes for depenedencies

2. The e2e tests are run with `oc cluster up` on travis:
   (i) scripts/oc-cluster.sh is used to bring up an Openshift cluster.
   (ii) scripts/travis-check-pods.sh is used to check if the router
        and docker-register pods are running once the oc cluster is
        up.

3. In `.travis.yml`, the e2e tests are run in parallel with the
   other tests

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -31,6 +31,17 @@ jobs:
...
         - bash <(curl -s https://codecov.io/bash)
 
+    - <<: *base-test
+      stage: test
+      go: 1.9
+      script:
+        - ./scripts/oc-cluster.sh
+        - make bin
+        - sudo cp ocdev /usr/bin
+        - oc login -u system:admin
```
Comment (kadel) [inline .travis.yml]: Is system admin really needed? 
We should use the regular user to run all the tests.

File: tests/e2e/e2e_test.go [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+	fmt.Print(string(out.Bytes()))
+	return string(out.Bytes())
+}
+
+func createApp(appName string) {
+	runCmd("ocdev application create " + appName)
+}
+
+func getApp() string {
+	app := runCmd("ocdev application get -q")
```
Comment (kadel) [inline tests/e2e/e2e_test.go]: lets use full flag name here `--short`

File: scripts/oc-cluster.sh [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+sudo service docker status
+
+## chmod needs sudo, so all other commands are with sudo
+sudo mkdir -p /home/travis/gopath/origin
+sudo chmod -R 766 /home/travis/gopath/origin
+
+## download oc binaries
+sudo wget https://github.com/openshift/origin/releases/download/v3.7.2/openshift-origin-client-tools-v3.7.2-282e43f-linux-64bit.tar.gz -O /home/travis/gopath/origin/openshift-origin-client-tools.tar.gz 2> /dev/null > /dev/null
+
+sudo wget https://github.com/openshift/origin/releases/download/v3.7.2/openshift-origin-server-v3.7.2-282e43f-linux-64bit.tar.gz  -O /home/travis/gopath/origin/openshift-origin-server.tar.gz 2> /dev/null > /dev/null
```
Comment (kadel) [inline scripts/oc-cluster.sh]: is sever used for something?

File: scripts/oc-cluster.sh [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+sudo service docker status
+
+## chmod needs sudo, so all other commands are with sudo
+sudo mkdir -p /home/travis/gopath/origin
+sudo chmod -R 766 /home/travis/gopath/origin
+
+## download oc binaries
+sudo wget https://github.com/openshift/origin/releases/download/v3.7.2/openshift-origin-client-tools-v3.7.2-282e43f-linux-64bit.tar.gz -O /home/travis/gopath/origin/openshift-origin-client-tools.tar.gz 2> /dev/null > /dev/null
+
+sudo tar -xvzf /home/travis/gopath/origin/openshift-origin-client-tools.tar.gz --strip-components=1 -C /bin
```
Comment (kadel) [inline scripts/oc-cluster.sh]: it might be better to use `/user/local/bin` instead of `/bin`

File: scripts/oc-cluster.sh [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+#export PATH=$PATH:/home/travis/gopath/origin/
+
+# Get oc version
+oc version
+
+# below cmd is important to get oc working in ubuntu
+sudo docker run -v /:/rootfs -ti --rm --entrypoint=/bin/bash --privileged openshift/origin:v3.7.2 -c "mv /rootfs/bin/findmnt /rootfs/bin/findmnt.backup"
```
Comment (kadel) [inline scripts/oc-cluster.sh]: what is this doing?

File: scripts/oc-cluster.sh [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+built=false
```
Comment (kadel) [inline scripts/oc-cluster.sh]: is this variable used somewhere?

File: scripts/oc-cluster.sh (hunk 5 above)
Comment (ashetty1) [reply scripts/oc-cluster.sh]: @kadel workaround for this: https://github.com/openshift/origin/issues/17017

File: .travis.yml [hunk 7]
```diff
@@ -31,6 +31,18 @@ jobs:
...
 
+    # Run e2e tests
+    - <<: *base-test
+      stage: test
+      go: 1.9
+      script:
+        - ./scripts/oc-cluster.sh
+        - make bin
+        - sudo cp ocdev /usr/bin
+        - go get -u github.com/onsi/ginkgo/ginkgo
```
Comment (kadel) [inline .travis.yml]: `ginkgo` is added vendored, we probably don't need to `go get` it here.

File: scripts/oc-cluster.sh [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+sudo service docker start
+sudo service docker status
+
+## chmod needs sudo, so all other commands are with sudo
+sudo mkdir -p /home/travis/gopath/origin
+sudo chmod -R 766 /home/travis/gopath/origin
+
+## download oc binaries
+sudo wget https://github.com/openshift/origin/releases/download/v3.7.2/openshift-origin-client-tools-v3.7.2-282e43f-linux-64bit.tar.gz -O /home/travis/gopath/origin/openshift-origin-client-tools.tar.gz 2> /dev/null > /dev/null
```
Comment (kadel) [inline scripts/oc-cluster.sh]: is there a reason why we downloading this to gopath?

File: Makefile [hunk 9]
```diff
@@ -66,6 +66,11 @@ prepare-release: cross
 test:
 	go test -race $(PKGS)
 
+# Run e2e tests
+.PHONY: test-e2e
+test-e2e:
+	go test -v github.com/redhat-developer/ocdev/tests/e2e
```
Comment (kadel) [inline Makefile]: do we really need to run it in verbose mode?

**Comment by cdrage** [comment]
Couldn't we simply use the e2e_test.go script that was written for Kedge for OCDEV?

Another thing too, here's the setup I used to get OpenShift running on SemaphoreCI (which I found way better since they have bigger instances)

You can access those settings here: https://semaphoreci.com/cdrage/kedge/settings (ignore the "cdrage" in the URl, you should still have access)

The scripts are as follows:

Setup:
```
# Free up some RAM
for s in apache2 cassandra elasticsearch memcached mongod postgresql sphinxsearch rabbitmq-server; do sudo service $s stop; done
# Remove unused languages
wget https://gist.githubusercontent.com/mimimalizam/e3fd7084f88384f4b2dfca035e3da2d4/raw/remove_unused_languages.sh && bash remove_unused_languages.sh
# Add SWAP
sudo swapoff -a
sudo dd if=/dev/zero of=/swapfile bs=1M count=1024
sudo mkswap /swapfile
sudo swapon /swapfile
# Download Minikube
curl -Lo minikube https://storage.googleapis.com/minikube/releases/v0.24.1/minikube-linux-amd64 && chmod +x minikube && sudo mv minikube /usr/local/bin/
# Download Kubectl
curl -LO https://storage.googleapis.com/kubernetes-release/release/v1.7.0/bin/linux/amd64/kubectl
chmod +x ./kubectl
sudo mv ./kubectl /usr/local/bin/kubectl
```

OpenShift:
```
# Install OC
export RELEASE="https://github.com/openshift/origin/releases/download/v3.6.0/openshift-origin-client-tools-v3.6.0-c4dd4cf-linux-64bit.tar.gz"
wget $RELEASE -O openshift.tar.gz
tar -xvf openshift.tar.gz --strip-components=1
sudo mv oc /usr/local/bin/oc
sudo chmod +x /usr/local/bin/oc
rm -r openshift.tar.gz README.md LICENSE
# Install minishift
sudo apt-get install -y virtualbox
export RELEASE="1.11.0"
wget https://github.com/minishift/minishift/releases/download/v$RELEASE/minishift-$RELEASE-linux-amd64.tgz -O minishift.tar.gz
tar -xvf minishift.tar.gz --strip-components=1
sudo mv minishift /usr/local/bin/minishift
rm README.adoc LICENSE
minishift start --vm-driver virtualbox
oc login -u system:admin
# Build Kedge
make bin
# Run tests
make test-e2e-os || true
```

Kubernetes:
```
# Setup Minikube with Docker (non-VM)
sudo minikube start --vm-driver=none
sudo mv /root/.kube $HOME/.kube || true
sudo chown -R $USER $HOME/.kube
sudo chgrp -R $USER $HOME/.kube
sudo mv /root/.minikube $HOME/.minikube || true
sudo chown -R $USER $HOME/.minikube
sudo chgrp -R $USER $HOME/.minikube 
# Build Kedge
make bin
# Run tests
VERBOSE=yes PARALLEL=4 TIMEOUT=50m make test-e2e
```

I believe it may be better to simply run these on Semaphore instead (since in my experience TravisCI kind of blows up in your face running Kubernetes / OpenShift clusters.. but that's just my opinion). 

@ashetty1 and @kadel I've sent you invites to the ocdev project I just setup.

I could help set this up these next few days since I've done it probably 5 times now with different projects, haha.

**Comment by cdrage** [comment]
Oh, and to note too. Instead of having these tests an extrodinarily long time (you mentioned that these tests are run in parallel on travis). Use Semaphore for e2e tests (like what we do with Kompose and Kedge) and use TravisCI for Unit Tests. Best of both worlds.

**Comment by ashetty1** [comment]
@cdrage super thanks. How about we have this travis thing going (It works and isn't taking a lot of time) while we get the semaphore functional? WDYT?

File: Makefile (hunk 9 above)
Comment (ashetty1) [reply Makefile]: Sure. WIll disable. Thought it would be useful to debug.

**Comment by cdrage** [comment]
@ashetty1 Sure, that works for me.

**Pull request merged**